src/ray/util/process.cc

```diff
@@ -242,21 +358,14 @@ class ProcessFD {
       }
       _exit(errno);  // fork() succeeded and exec() failed, so abort the child
     }
-    if (pid > 0) {
-      // Parent process case
-      if (decouple) {
-        int s;
-        (void)waitpid(pid, &s, 0);  // can't do much if this fails, so ignore return value
-        int r = read(pipefds[0], &pid, sizeof(pid));
-        (void)r;  // can't do much if this fails, so ignore return value
-      }
-    }
+    // Parent process case
     // Use pipe to track process lifetime. (The pipe closes when process terminates.)
     fd = pipefds[0];
     if (pid == -1) {
       ec = std::error_code(errno, std::system_category());
     }
 #endif
+    addOwnedChild(pid);
```

Comment:
You need a signal block between process creation and adding it to the owned set. Otherwise there is a race where the process dies before it is added to the owned set. I'll leave to you whether this is something that should be fixed now or not.

Reply:
(maybe boost magic solves this for you)